colexec: increase testing of joiners

yuzefovich opened the pull request:
**colexec: add unit tests with ON expressions for joiners**

This commit adds unit tests for hash and merge joiners with ON
expression (for join types that are currently supported). Both
tests are refactored to use NewColOperator instead of creating the
operator directly which is needed because ON expression with INNER
join is handled as a filter on top of the joiner.

Fixes: #40634.

**colexec: add post filter when testing joiners against the processors**

This commit adds a filter to the post processing step of testing
hash and merge join operators against the corresponding processors.

Release note: None

yuzefovich committed:
colexec: add unit tests with ON expressions for joiners

This commit adds unit tests for hash and merge joiners with ON
expression (for join types that are currently supported). Both
tests are refactored to use NewColOperator instead of creating the
operator directly which is needed because ON expression with INNER
join is handled as a filter on top of the joiner.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41038)
<!-- Reviewable:end -->

yuzefovich commented:
First three commits are a part of the separate PR, so please look only at the fourth commit here.

yuzefovich committed:
colexec: add post filter when testing joiners against the processors

This commit adds a filter to the post processing step of testing
hash and merge join operators against the corresponding processors.

Release note: None

yuzefovich commented:
This is now RFAL.

yuzefovich commented:
Friendly ping.

asubiotto reviewed (commented):
> Reviewed 5 of 5 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41038)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/hashjoiner_test.go, line 815 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrAZgFa791kHmi6DdCz:-LrAZgFb519l55t5B40K:baq997h) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/hashjoiner_test.go#L815)):*
> ```Go
> }
> 
> func createSpecForHashJoiner(tc hjTestCase) *execinfrapb.ProcessorSpec {
> ```

I like this. The `rowexec` tests were essentially creating specs straight up, putting too much complexity on writing a test case, but this conversion step fixes that.
___
*[pkg/sql/colexec/mergejoiner_test.go, line 1578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrA_ejvDQm3YF1XVKwV:-LrA_ejw-DHQHAbzGMC4:bvd4szd) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1578)):*
> ```Go
> 				// When we have an inner join with ON expression, a filter operator
> 				// will be put on top of the merge join, so to make life easier, we'll
> 				// just ignore the requested output batch size.
> ```

Are we losing test coverage by not doing this?
___
*[pkg/sql/colexec/mergejoiner_test.go, line 1601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrAa9tF9PDLlatPSsnw:-LrAa9tF9PDLlatPSsnx:b2s52xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1601)):*
> ```Go
> 		runner(t, []tuples{tc.leftTuples, tc.rightTuples}, nil /* typs */, tc.expected, mergeJoinVerifier,
> 			func(input []Operator) (Operator, error) {
> 				lOrderingCols := make([]execinfrapb.Ordering_Column, len(tc.leftEqCols))
> ```

Might be nice to extract all these things into a little function like you did with the hash joiner.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41038)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/mergejoiner_test.go, line 1578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrA_ejvDQm3YF1XVKwV:-LrB4rMLBY3ea4Ap9ZeM:b8gvme9) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1578)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Are we losing test coverage by not doing this?
</blockquote></details>

Technically speaking, yes, but not really. This occurs only when we have an ON expression, but all test cases for INNER join (even those with ON expression) run without the ON expression, so the requested output batch size will still be exercised. Only cases that are "lost" by this change is when we choose to run INNER join *with* ON expression. 
___
*[pkg/sql/colexec/mergejoiner_test.go, line 1601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrAa9tF9PDLlatPSsnw:-LrB6oRQ9qOUOYdPaYQa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1601)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Might be nice to extract all these things into a little function like you did with the hash joiner.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41038)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/mergejoiner_test.go, line 1578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrA_ejvDQm3YF1XVKwV:-LrB8Dik7o0U6yowwp2p:byzpzr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1578)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Technically speaking, yes, but not really. This occurs only when we have an ON expression, but all test cases for INNER join (even those with ON expression) run without the ON expression, so the requested output batch size will still be exercised. Only cases that are "lost" by this change is when we choose to run INNER join *with* ON expression. 
</blockquote></details>

To clarify: the test cases that have ON expression are run both without and with ON expression, and the former still exercises this path.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> minus the small testing concern

> Reviewed 3 of 3 files at r3, 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-:-LrPLGiDC4jyuFOzSHI0:bns543t)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/mergejoiner_test.go, line 1578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrA_ejvDQm3YF1XVKwV:-LrPJKB_E8MAlXjIK575:b-4e311x) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1578)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

To clarify: the test cases that have ON expression are run both without and with ON expression, and the former still exercises this path.
</blockquote></details>

I think it would be good to also test the handling of batches with an `ON` expression tacked on after the joiner (or some unit tests ensuring the `ON` pipeline manages batches correctly), but I'll leave it up to you.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

I think the second commit addresses your concern, and if that is not enough, we'll add more tests later.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-:-LrPdwoP2y0_miEL-BKR:bf3slu3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/mergejoiner_test.go, line 1578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41038#-LrA_ejvDQm3YF1XVKwV:-LrPckVhB37gD-YyIZSO:bmn55nl) ([raw file](https://github.com/cockroachdb/cockroach/blob/817cb5c80a21a3854c60e71b1889ec4b065fbd9c/pkg/sql/colexec/mergejoiner_test.go#L1578)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it would be good to also test the handling of batches with an `ON` expression tacked on after the joiner (or some unit tests ensuring the `ON` pipeline manages batches correctly), but I'll leave it up to you.
</blockquote></details>

So that is done in the second commit in this PR: with `Test(Hash|Merge)JoinerAgainstProcessor` we will try adding a post filter after the joiner in every second try, and for all join types that support `ON` expression we also exercise that in there, so we will have test configurations that use both `ON` expression on the joiner and a post filter. I think it is sufficient.

I think of tests in `hashjoiner_test.go` and `mergejoiner_test.go` as unit tests - they should be testing one thing, and of `Test(Hash|Merge)JoinerAgainstProcessor` as integration tests - they should be testing many things at once.


<!-- Sent from Reviewable.io -->
